### DISCUSSION TITLE
Missing bootable label when encapsulating a commit

### DISCUSSION THREAD

**Discussion by cgwalters** [ostreedev/ostree#2978] [Q&A] [answered]
Trying to create an ostree container out of a commit, and the resulting image does not have a ostree.bootable label. When I try to import the container with ostree container unencapsulate I run into the error added in this commit: https://github.com/ostreedev/ostree-rs-ext/commit/9c4a75b3778a3f2fdece095f8f5f7a6289ab512dLooks like the purpose of this condition was to identify ostree container images, but in my case the image did not get this label, and it is an ostree image. Is this a bug? Or am I misunderstanding?

**Comment by cgwalters** [comment]
As a workaround we've added the `--label ostree.bootable=true` argument to the encapsulate command and that seems to allow us to import our commit

**Reply by cgwalters** [reply]
How are you generating the original commit?  Nowadays you want to use `ostree commit --bootable`, which is what rpm-ostree does by default for example.

The encapsulation process will automatically preserve that metadata.

**Reply by omertuc** [reply]
In our case, our commit doesn't contain a bootable kernel, so it doesn't really make sense for us to add that flag

>  --bootable                                       Flag this commit as a bootable OSTree (e.g. contains a Linux kernel)

**Reply by omertuc** [reply]
We are in essence slightly abusing ostree as a file-delivery mechanism. Maybe it is the intent of ostree to have a bootable kernel in every commit, but in our case we're just storing 2 tarballs and a metadata file inside a commit.

We can mark the commit as `--bootable`, or add a fake `--label ostree.bootable=true` during encapsulation, but it feels hacky

**Reply by cgwalters** [reply]
Ohhh sorry!  I misunderstood.   Wait but...I only see the `container store` code processing this.

Precisely what commands are you running here?  Are you sure it's `ostree container unencapsulate` that's failing?

ostree is absolutely intended to also support non-bootable things to be clear, but in practice...that's kind of our focus.

**Reply by javipolo** [reply]
I'm testing with the version:
```
[root@hub-2 ~]# ostree --version
libostree:
 Version: '2022.2'
 Git: a332611316fee36b1c630f8b266e4a37f77f2c54
 Features:
  - libcurl
  - libsoup
  - gpgme
  - libarchive
  - selinux
  - openssl
  - libmount
  - systemd
  - release
  - p2p
```

# Expected way, without bootable label
On one host we create a commit, and with that an OCI container:
```
[root@hub-1 core]# ostree container encapsulate test registry:quay.io/javipolo/ostmagic:test --repo /ostree/repo                                                                                 
sha256:cb1394ced4b345b12c695f420bdd936ab94ec08e5f030d630c787a84a2d2f3c2
```

On another host we try to unencapsulate that:
```
[root@hub-2 ~]# ostree container unencapsulate --repo /ostree/repo ostree-unverified-registry:quay.io/javipolo/ostmagic:test                                                                     
   Downloading...                                                                                                                                                                             
error: Fetching manifest: Target image does not have ostree.bootable label
```

# Adding the label as a workaround
However, if we add the ostree.bootable=true label, we can unencapsulate without problems:

```
[root@hub-1 core]# ostree container encapsulate test registry:quay.io/javipolo/ostmagic:test2 --repo /ostree/repo --label ostree.bootable=true                                                   
sha256:3314c030cc12f265cd768967b57644c746fb1288f50c1460da756229ef779126
```

```
[root@hub-2 ~]# ostree container unencapsulate --repo /ostree/repo ostree-unverified-registry:quay.io/javipolo/ostmagic:test2
   Downloading...                                                                                                                                                                              warning: Image is using v0 export layout, this is deprecated and support will be dropped in the future
Imported: ac1245f4abb9ff9f156bdf0559bbf05bd1ed8da43ad02b2ecb519eb89d9e9702

```

**Reply by omertuc** [reply]
>I only see the container store code processing this

unencapsulate CLI [calls](https://github.com/ostreedev/ostree-rs-ext/blob/ea853c082c09594e1559e55b91d011994feb855b/lib/src/cli.rs#L835) `container_import` [calls](https://github.com/ostreedev/ostree-rs-ext/blob/ea853c082c09594e1559e55b91d011994feb855b/lib/src/cli.rs#L588) `unencapsulate` [calls](https://github.com/ostreedev/ostree-rs-ext/blob/ea853c082c09594e1559e55b91d011994feb855b/lib/src/container/store.rs#L706) `prepare_internal` [calls](https://github.com/ostreedev/ostree-rs-ext/blob/ea853c082c09594e1559e55b91d011994feb855b/lib/src/container/store.rs#L560) `parse_manifest_layout` which [errors out](https://github.com/ostreedev/ostree-rs-ext/blob/ea853c082c09594e1559e55b91d011994feb855b/lib/src/container/store.rs#L355)

**Reply by cgwalters** [reply, marked as answer]
Yes, you're absolutely right.  PR to fix this in https://github.com/ostreedev/ostree-rs-ext/pull/513
